sklearn/tree/_export.py

```diff
@@ -953,16 +953,16 @@ def export_text(
         It can be an instance of
         DecisionTreeClassifier or DecisionTreeRegressor.
 
-    feature_names : list of str, default=None
-        A list of length n_features containing the feature names.
+    feature_names : array-like of shape (n_features,), default=None
+        An array containing the feature names.
         If None generic names will be used ("feature_0", "feature_1", ...).
 
-    class_names : list or None, default=None
+    class_names : array-like or None, default=None
```

Comment:
An "array-like" could be a custom object that only defines `__array__`:

```python
import numpy as np

class MyArrayLike:
    def __init__(self, ndarray):
        self._inner_ndarray = ndarray
        
    def __array__(self):
        return self._inner_ndarray
    
class_names = MyArrayLike(np.array(["a", "b"]))

try:
    len(classs_names)
except TypeError as e:
    print(e)
# object of type 'MyArrayLike' has no len()

try:
    classs_names[0]
except TypeError as e:
    print(e)
# 'MyArrayLike' object is not subscriptable


```

To support fully array-likes, we'll need to call `asarray` on it before operating on it:

```python
class_names = MyArrayLike(np.array(["a", "b"]))
class_names_array = np.asarray(class_names)

print(len(class_names_array))
print(class_names_array[0])
```

Note the downside of calling `asarray` is that it will make a copy when the input is a list.

### Alternative

`export_*` only requires the input object to define `__len__` and a  `__getitem__` that returns a string. This interface is provided by `list` and `ndarray`, but not custom objects that only define `__array__` to be "array-like". For this PR, I'll prefer to be strict here and accept "list" and "ndarary" explicit and not the generic "array-like". Concretely:

```
class_names : list, ndarray, or None, default=None
```